### ISSUE TITLE
Tendermint admission controls based on cosmic-swingset state

### ISSUE THREAD

**Issue by mhofman** [Agoric/Agoric#5176]
## What is the Problem Being Solved?

To support #5334 ~~and more specifically #5027~~, we need the ability to only allow some messages into the mempool and into blocks based on the system's state:
- reject messages from getting into the mempool/blocks if Swingset is overloaded.
- accept messages from ~~high priority~~ economy critical senders if ~~low-priority~~ non-economic message are currently rejected
- accept cosmos governance messages that update the admission policy itself

## Description of the Design

The JavaScript side of cosmic-swingset will store the admission policy and expose it and the current state of Swingset (#5350). This info will likely be returned as the result of the init and commit actions and should be cached at the cosmos level for subsequent usage.

Policy updates would come in the form of cosmos level governance messages and would be routed to cosmic-swingset in a dedicated action. The admission checks should probably hard code these type of governance transactions as always being allowed regardless of policy so that the chain cannot permanently DOS itself. See https://github.com/Agoric/agoric-sdk/issues/5335

The info returned by cosmic-swingset is guaranteed to be deterministic and can be used for consensus block checks. Of course it can also be used for admissibility checks into the mempool.

Some potential pointers:
[`CheckAdmissibility`](https://github.com/Agoric/agoric-sdk/blob/adfdf3638f56e8f37d34661f033f42843354db26/golang/cosmos/x/swingset/types/msgs.go#L31) may be capable of this, while having access to current state.

There may be some further information the [Cosmos docs for transaction lifecycle](https://github.com/cosmos/cosmos-sdk/blob/58597139fa0fb9e9be60deebee3df1663aa2cfaf/docs/basics/tx-lifecycle.md).

If tendermint doesn't provide context for the mempool admissibility checks, the admission policy state may need to be store globally...

## Security Considerations
This mechanism can in effect cause a pause of all swingset ingress, including economic governance. At this point it cannot pause actions that have already been delivered into swingset from progressing internally.

## Test Plan
TBD

**Comment by mhofman** [comment]
#5281 will add a `controller.getQueuesLength()` which returns 2 numbers: an `activeQueues` and an `inactiveQueues`. The former represent queues that can be processed next without any further state change. The latter represents messages pending on some state change (in particular a promise resolution, or pipeline), which could happen after some other delivery is made. The sum of both is probably the number we want to consider as the "total queue depth" but if there is nothing in the activeQueues, messages need to be delivered into swingset to nudge it.

Another thing that came up is that we may want to have an hysteresis for the queue depth since the number of messages in the different queues is not expected to be constant crank after crank (e.g. one delivery can cause a bunch of new queued events, which then coalesces back down over a few cranks)

**Comment by mhofman** [comment]
The exact content of the admission policy is TBD. I'm inclined to have the cosmos relevant parts opaque to cosmic-swingset. It should probably include:
- A list of message types considered to be part of the economy (cosmos relevant)
- A list of senders allowed to send these messages (cosmos relevant)
- Either a boolean indicating whether the queue based conditions were tripped, or the raw data to let cosmos compute this. The argument for the latter would be to limit the number of new messages allowed in when Swingset queues are close to being full.

**Comment by Tartuffo** [comment]
@mhofman / @warner Is this issue (and its blocker) a dup with our other backpressure issues?  Or part of them (not part of the epic).

**Comment by mhofman** [comment]
We met with @JimLarson and @arirubinstein and some things came up.

Ari recommended against having mempool inclusion based on some local state (aka which/how many tx have been added to the mempool already) because it would result in validator mempools diverging. That's apparently bad because when a block is proposed, validators that have not heard of all transactions in the block will not be able to vote on the block, potentially causing churn on block proposers. If we were to go that way, Jim suggested a green/yellow/red transaction categorization in an AnteHandler based on local state, and have orange+red rejected for CheckTx and red rejected for DeliverTx.

The next best thing would be a filtering function on inclusion from the mempool into a block, which also can be local/non-deterministic, but wouldn't be subject to the block voting issue. This could potentially be built on top of the [mempool prioritization mechanism](https://medium.com/tendermint/tendermint-v0-35-introduces-prioritized-mempool-a-makeover-to-the-peer-to-peer-network-more-61eea6ec572d) in newer cosmos SDK. Unless some other mechanism controlled the mempool acceptance, this would not provide actual back-pressure.

Finally the last solution is for swingset to queue transactions internally and only process them once its main queues are back under the thresholds. The main drawback from this approach is that SwingSet needs to learn about sender priorities before priorities are actually implemented and that this queuing would also not provide actual back-pressure.

**Comment by JimLarson** [comment]
One issue with using the AnteHandler at CheckTx time for backpressure is that CheckTx uses the most recent committed state, and thus may vary as the candidate Tx is gossipped around the various mempools. If I understand this correctly, we already deal with mixed mempool acceptance/rejection from different validators, e.g. if an account balance dips below the ability to pay gas fees at height N, a near-simultaneous Tx might be accepted into the mempool at validators where it arrives before height N and rejected where it arrives at N or later. A proposer which accepted and tries to include it in a block would be rightly miffed to have its proposed block rejected by the others. Hopefully there's a way for such a validator to purge that Tx from its mempool.

This is probably rare and relatively unimportant when it affects a single account, but it could be more of a problem for the chain if we suddenly shut the door on all non-prioritized messages to Swingset.

It might be better to just have a simpler single threshold for rejecting such messages at both CheckTx and DeliverTx time, so that a validator which accepted a message based on old state has a better changes of purging the message based on updated state when it tries to include it in a block. @arirubinstein let me know if I've misunderstood the lifecycle.

**Comment by mhofman** [comment]
Given it's not clear what may be the impact of having inconsistent mempool state between validators, we decided to deprioritize this and focus on a solution based on priority queues at the cosmic-swingset layer. However for real backpressure, we do need to figure out if a mempool based approach is possible, and the potential impacts on Tendermint.

**Comment by mhofman** [comment]
This was in effect duplicated by https://github.com/Agoric/agoric-sdk/issues/6314 and implemented in https://github.com/Agoric/agoric-sdk/pull/6381
